iotdb-core/ainode/ainode/core/inference/inference_request_pool.py

```diff
@@ -65,65 +67,88 @@ def __init__(
+        memory_before_warmup = torch.cuda.memory_allocated(self.device)
+        logger.info(
+            f"Before warm-up, peak memory usage: {memory_before_warmup:.2f} bytes"
+        )
 
-    def _activate_requests(self):
-        if self._waiting_queue.empty():
-            return
-        request: InferenceRequest = self._waiting_queue.get()
-        # TODO: Check memory size before activating requests
-        request.inputs = request.inference_pipeline.preprocess_inputs(request.inputs)
-        request.mark_running()
-        logger.debug(
-            f"[Inference][Device-{self.device}][Pool-{self.pool_id}][ID-{request.req_id}] Request is activated with inputs shape {request.inputs.shape}"
+        # warm-up
+        with torch.no_grad():
+            self.model.generate(dummy_input, max_new_tokens=1)
+        torch.cuda.synchronize(self.device)
+        peak_memory_1_token = torch.cuda.max_memory_allocated(self.device)
+        logger.info(
+            f"Baseline memory usage for 1 token: {peak_memory_1_token:.2f} bytes"
+        )
+        logger.info(
+            f"Differentiation : {peak_memory_1_token-memory_before_warmup:.2f} bytes"
         )
-        self._running_queue.put(request)
+
+    def _activate_requests(self):
+        requests = self._scheduler.schedule_activate()
+        for request in requests:
+            request.inputs = request.inference_pipeline.preprocess_inputs(
+                request.inputs
+            )
+            logger.debug(
+                f"[Inference][Device-{self.device}][Pool-{self.pool_id}][ID-{request.req_id}] Request is activated with inputs shape {request.inputs.shape}"
+            )
 
     def _requests_activate_loop(self):
         while not self._stop_event.is_set():
             time.sleep(self.WAITING_INTERVAL_IN_MS / 1000)
             self._activate_requests()
 
     def _step(self):
-        if self._running_queue.empty():
-            return
+        requests = self._scheduler.schedule_step()
         # TODO: We need a batcher to accelerate the concurrent inference
-        # TODO: Check memory size before executing requests
-        request: InferenceRequest = self._running_queue.get()
-        inputs = request.inputs.to(self.device)
-        output = self.model.generate(
-            inputs,
-            max_new_tokens=request.max_new_tokens,
-            num_samples=10,
-            revin=True,
-        )
-        request.output_tensor = request.output_tensor.to(
-            self.device
-        )  # Ensure output tensor is on the same device
-        request.write_step_output(output[0].mean(dim=0))
-        request.inference_pipeline.post_decode()
-        if request.is_finished():
-            request.inference_pipeline.post_inference()
-            logger.debug(
-                f"[Inference][Device-{self.device}][Pool-{self.pool_id}][ID-{request.req_id}] Request is finished"
+        for request in requests:
+            inputs = request.inputs.to(self.device)
```

Comment:
Be cautious to this line, seems it is redundant when the `step` is executed for multiple times.